MaxText/maxtext_utils.py

```diff
@@ -124,3 +125,39 @@ def calculate_tflops_prefill(num_model_parameters, prefill_length, config, log=T
           f'\t\tCausal attention TFLOPs: {causal_attention_tflops} ',
           f'({100 * causal_attention_tflops/total_tflops:.2f})% of Total')
   return total_tflops, learnable_weight_tflops, causal_attention_tflops
+
+
+def assert_params_sufficiently_sharded(params, mesh, tolerance=0.01):
+  """Checks whether most params are sharded across sharding axis.
+
+  This function determines whether the majority of parameters  are distributed
+  across a specified sharding axes with an acceptable tolerance. It compares the
+  current distribution to a scenario where all parameters are fully sharded
+  across the 'fsdp', 'fsdp_transpose', 'sequence', and 'tensor' axes.
+
+  Args:
+    params: params of the model state
+    mesh: mesh constructed from config
+    tolerance: float between 0.0 and 1.0 representing the allowed percentage of
+    non-sharded parameters.
+  Returns:
+    bool: True if the majority of parameters are sufficiently sharded
+  """
+  total_num_params = max_utils.calculate_num_params_from_pytree(params) * 1e9
+  product_num_devices_for_weight_sharding  = 1
+  for axis in ['fsdp', 'fsdp_transpose','sequence', 'tensor']:
+    product_num_devices_for_weight_sharding *= mesh.shape[axis]
```

Comment:
how does this work? Isn't mesh an array? Is `mesh.shape` actually a dictionary? Lots I don't understand here.

Reply:
mesh.shape is an OrderedDict. E.g. 
```
(Pdb) mesh.shape
OrderedDict([('data', 1), ('fsdp', 1), ('fsdp_transpose', 1), ('sequence', 1), ('tensor', 4), ('autoregressive', 1)])
```

Reply:
Mesh is `jax.sharding.Mesh` which has a method shape which returns an OrderedDict

Reply:
amazing, I had no idea.